File: typescript/src/parsers/item-extractor.ts
/**
 * SEC Filing Item Extractor
 * 
 * Extracts individual items from SEC filings (10-K, 10-Q, 8-K, etc.)
 * based on standard item numbering and structure.
 */

import { InvalidFormTypeError } from '../exceptions/errors';

export enum FormType {
  FORM_10K = "10-K",
  FORM_10Q = "10-Q",
  FORM_8K = "8-K",
  FORM_20F = "20-F",
  FORM_40F = "40-F",
}

export interface ItemDefinition {
  number: string;
  title: string;
  aliases?: string[];
  required?: boolean;
}

export interface ExtractedItem {
  itemNumber: string;
  title: string;
  content: string;
  startPosition: number;
  endPosition: number;
}

export class ItemExtractor {
  // 10-K Item definitions
  private static readonly FORM_10K_ITEMS: ItemDefinition[] = [
    { number: "1", title: "Business" },
    { number: "1A", title: "Risk Factors" },
    { number: "1B", title: "Unresolved Staff Comments" },
    { number: "1C", title: "Cybersecurity", required: false },
    { number: "2", title: "Properties" },
    { number: "3", title: "Legal Proceedings" },
    { number: "4", title: "Mine Safety Disclosures", required: false },
    { number: "5", title: "Market for Registrant's Common Equity" },
    { number: "6", title: "Reserved", required: false },
    { number: "7", title: "Management's Discussion and Analysis", aliases: ["MD&A"] },
    { number: "7A", title: "Quantitative and Qualitative Disclosures About Market Risk" },
    { number: "8", title: "Financial Statements and Supplementary Data" },
    { number: "9", title: "Changes in and Disagreements with Accountants" },
    { number: "9A", title: "Controls and Procedures" },
    { number: "9B", title: "Other Information" },
    { number: "9C", title: "Disclosure Regarding Foreign Jurisdictions", required: false },
    { number: "10", title: "Directors, Executive Officers and Corporate Governance" },
    { number: "11", title: "Executive Compensation" },
    { number: "12", title: "Security Ownership" },
    { number: "13", title: "Certain Relationships and Related Transactions" },
    { number: "14", title: "Principal Accountant Fees and Services" },
    { number: "15", title: "Exhibits and Financial Statement Schedules" },
  ];

  // 10-Q Item definitions
  private static readonly FORM_10Q_ITEMS: ItemDefinition[] = [
    { number: "1", title: "Financial Statements" },
    { number: "2", title: "Management's Discussion and Analysis", aliases: ["MD&A"] },
    { number: "3", title: "Quantitative and Qualitative Disclosures About Market Risk" },
    { number: "4", title: "Controls and Procedures" },
    { number: "1", title: "Legal Proceedings", aliases: ["Part II, Item 1"] },
    { number: "1A", title: "Risk Factors", aliases: ["Part II, Item 1A"] },
    { number: "2", title: "Unregistered Sales of Equity Securities", aliases: ["Part II, Item 2"] },
    { number: "3", title: "Defaults Upon Senior Securities", aliases: ["Part II, Item 3"] },
    { number: "4", title: "Mine Safety Disclosures", aliases: ["Part II, Item 4"], required: false },
    { number: "5", title: "Other Information", aliases: ["Part II, Item 5"] },
    { number: "6", title: "Exhibits", aliases: ["Part II, Item 6"] },
  ];

  // 8-K Item definitions
  private static readonly FORM_8K_ITEMS: ItemDefinition[] = [
    { number: "1.01", title: "Entry into a Material Definitive Agreement" },
    { number: "1.02", title: "Termination of a Material Definitive Agreement" },
    { number: "2.01", title: "Completion of Acquisition or Disposition of Assets" },
    { number: "2.02", title: "Results of Operations and Financial Condition" },
    { number: "2.03", title: "Creation of a Direct Financial Obligation" },
    { number: "3.01", title: "Notice of Delisting or Failure to Satisfy" },
    { number: "3.02", title: "Unregistered Sales of Equity Securities" },
    { number: "4.01", title: "Changes in Registrant's Certifying Accountant" },
    { number: "4.02", title: "Non-Reliance on Previously Issued Financial Statements" },
    { number: "5.01", title: "Changes in Control of Registrant" },
    { number: "5.02", title: "Departure of Directors or Certain Officers" },
    { number: "5.03", title: "Amendments to Articles of Incorporation or Bylaws" },
    { number: "7.01", title: "Regulation FD Disclosure" },
    { number: "8.01", title: "Other Events" },
    { number: "9.01", title: "Financial Statements and Exhibits" },
  ];

  private readonly formItems: Map<FormType, ItemDefinition[]>;

  constructor() {
    this.formItems = new Map([
      [FormType.FORM_10K, ItemExtractor.FORM_10K_ITEMS],
      [FormType.FORM_10Q, ItemExtractor.FORM_10Q_ITEMS],
      [FormType.FORM_8K, ItemExtractor.FORM_8K_ITEMS],
    ]);
  }

  /**
   * Extract all items from a filing
   * @param content The filing content (HTML or text)
   * @param formType The type of form (e.g., "10-K", "10-Q", "8-K")
   * @returns Dictionary mapping item numbers to their content
   */
  extractItems(content: string, formType: string | FormType): Record<string, string> {
    // Convert string form type to enum
    const parsedFormType = typeof formType === 'string' ? this.parseFormType(formType) : formType;

    if (!this.formItems.has(parsedFormType)) {
      throw new InvalidFormTypeError(formType.toString(), ['10-K', '10-Q', '8-K', '20-F', '40-F']);
    }

    // Clean content
    const cleanContent = this.cleanContent(content);

    // Extract table of contents if available
    const tocItems = this.extractTableOfContents(cleanContent);

    // Extract items
    const items = this.extractItemsFromContent(cleanContent, parsedFormType, tocItems);

    // Post-process and validate
    return this.postProcessItems(items, parsedFormType);
  }

  /**
   * Extract specific items from a filing
   * @param content The filing content
   * @param formType The type of form
   * @param itemNumbers List of item numbers to extract
   * @returns Dictionary with only the requested items
   */
  extractSpecificItems(
    content: string,
    formType: string | FormType,
    itemNumbers: string[]
  ): Record<string, string> {
    const allItems = this.extractItems(content, formType);
    const result: Record<string, string> = {};
    
    for (const itemNum of itemNumbers) {
      if (itemNum in allItems) {
        result[itemNum] = allItems[itemNum];
      }
    }
    
    return result;
  }

  /**
   * Get item definitions for a specific form type
   * @param formType The form type
   * @returns List of item definitions
   */
  getItemDefinitions(formType: string | FormType): ItemDefinition[] {
    const parsedFormType = typeof formType === 'string' ? this.parseFormType(formType) : formType;
    return this.formItems.get(parsedFormType) || [];
  }

  private parseFormType(formTypeStr: string): FormType {
    const upperType = formTypeStr.toUpperCase();

    if (upperType.includes("10-K") || upperType.includes("10K")) {
      return FormType.FORM_10K;
    } else if (upperType.includes("10-Q") || upperType.includes("10Q")) {
      return FormType.FORM_10Q;
    } else if (upperType.includes("8-K") || upperType.includes("8K")) {
      return FormType.FORM_8K;
    } else if (upperType.includes("20-F") || upperType.includes("20F")) {
      return FormType.FORM_20F;
    } else if (upperType.includes("40-F") || upperType.includes("40F")) {
      return FormType.FORM_40F;
    } else {
      throw new InvalidFormTypeError(formTypeStr, ['10-K', '10-Q', '8-K', '20-F', '40-F']);
    }
  }

  private cleanContent(content: string): string {
    // Remove HTML tags but preserve structure
    let cleaned = content.replace(/<[^>]+>/g, ' ');

    // Normalize whitespace
    cleaned = cleaned.replace(/\s+/g, ' ');

    // Preserve line breaks for item boundaries
    cleaned = cleaned.replace(/(Item\s+\d+[A-Z]?\.)/gi, '\n\n$1');

    return cleaned.trim();
  }

  private extractTableOfContents(content: string): Array<[string, number]> {
    const tocItems: Array<[string, number]> = [];

    // Look for table of contents section
    const tocMatch = content.match(
      /TABLE\s+OF\s+CONTENTS(.*?)(?:Item\s+1\.|PART\s+I\s)/is
    );

    if (tocMatch) {
      const tocContent = tocMatch[1];

      // Extract item references from TOC
      const itemPattern = /Item\s+(\d+[A-Z]?)\.\s*([^\n\r.]+)/gi;
      let match;

      while ((match = itemPattern.exec(tocContent)) !== null) {
        const itemNum = match[1].toUpperCase();
        tocItems.push([itemNum, match.index]);
      }
    }

    return tocItems;
  }

  private extractItemsFromContent(
    content: string,
    formType: FormType,
    tocItems: Array<[string, number]>
  ): Map<string, ExtractedItem> {
    const items = new Map<string, ExtractedItem>();
    const itemDefinitions = this.formItems.get(formType) || [];

    for (const itemDef of itemDefinitions) {
      // Build patterns for each item
      const patterns: string[] = [
        `Item\\s+${this.escapeRegex(itemDef.number)}\\.\\s*${this.escapeRegex(itemDef.title)}`,
        `Item\\s+${this.escapeRegex(itemDef.number)}\\.\\s*(?=[A-Z])`,
        `Item\\s+${this.escapeRegex(itemDef.number)}(?:\\.|:|\\s)`,
      ];

      // Add alias patterns
      if (itemDef.aliases) {
        for (const alias of itemDef.aliases) {
          patterns.push(this.escapeRegex(alias));
        }
      }

      // Try each pattern
      for (const pattern of patterns) {
        const regex = new RegExp(pattern, 'gi');
        const matches = Array.from(content.matchAll(regex));

        if (matches.length > 0) {
          // Use the first match after TOC (if TOC exists)
          let match = matches[0];
          
          if (matches.length > 1 && tocItems.length > 0) {
            // Skip matches that appear in TOC
            for (const m of matches.slice(1)) {
              if (!this.isInToc(m.index!, tocItems)) {
                match = m;
                break;
              }
            }
          }

          const startPos = match.index!;

          // Find the end position (start of next item)
          const endPos = this.findItemEnd(content, startPos, itemDefinitions);

          // Extract content
          const itemContent = content.substring(startPos, endPos).trim();

          items.set(itemDef.number, {
            itemNumber: itemDef.number,
            title: itemDef.title,
            content: itemContent,
            startPosition: startPos,
            endPosition: endPos,
          });
          
          break;
        }
      }
    }

    return items;
  }

  private isInToc(position: number, tocItems: Array<[string, number]>): boolean {
    if (tocItems.length === 0) return false;

    // Rough heuristic: if position is before the last TOC item + buffer
    const lastTocPos = Math.max(...tocItems.map(item => item[1]));
    return position < lastTocPos + 500;
  }

  private findItemEnd(
    content: string,
    startPos: number,
    _itemDefinitions: ItemDefinition[]
  ): number {
    // Look for the next item
    const nextItemPattern = /Item\s+\d+[A-Z]?[.:]\s*[A-Z]/i;
    const remainingContent = content.substring(startPos + 10);
    const match = remainingContent.match(nextItemPattern);

    if (match && match.index !== undefined) {
      return startPos + 10 + match.index;
    } else {
      // No next item found, return end of content
      return content.length;
    }
  }

  private postProcessItems(
    items: Map<string, ExtractedItem>,
    _formType: FormType
  ): Record<string, string> {
    const processed: Record<string, string> = {};

    for (const [itemNum, extractedItem] of items) {
      // Clean up the content
      let content = extractedItem.content;

      // Remove excessive whitespace
      content = content.replace(/\n\s*\n\s*\n/g, '\n\n');

      // Ensure we have some content
      if (content.trim().length > 50) {
        processed[itemNum] = content;
      } else {
        // Try to handle empty or placeholder items
        if (content.toLowerCase().includes("none") || 
            content.toLowerCase().includes("not applicable")) {
          processed[itemNum] = content;
        } else {
          processed[itemNum] = "";
        }
      }
    }

    return processed;
  }

  private escapeRegex(str: string): string {
    return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  }
}

File: python/src/sec_edgar_toolkit/parsers/item_extractor.py
"""
SEC Filing Item Extractor

Extraction engine that pulls individual items out of SEC filings
(10-K, 10-Q, 8-K) using the definitions in :mod:`.items`.
"""

import re
from typing import Dict, List, Optional, Tuple, Union

from .items import (
    FORM_8K_ITEMS,
    FORM_10K_ITEMS,
    FORM_10Q_ITEMS,
    FORM_ITEM_DEFINITIONS,
    EightKItem,
    ExtractedItem,
    FormType,
    ItemDefinition,
    TenKItem,
    TenQItem,
)

__all__ = [
    "ItemExtractor",
    "FormType",
    "ItemDefinition",
    "ExtractedItem",
    "TenKItem",
    "TenQItem",
    "EightKItem",
    "FORM_10K_ITEMS",
    "FORM_10Q_ITEMS",
    "FORM_8K_ITEMS",
    "FORM_ITEM_DEFINITIONS",
]


class ItemExtractor:
    """Extracts individual items from SEC filings."""

    FORM_10K_ITEMS = FORM_10K_ITEMS
    FORM_10Q_ITEMS = FORM_10Q_ITEMS
    FORM_8K_ITEMS = FORM_8K_ITEMS

    def __init__(self):
        """Initialize the item extractor."""
        self.form_items = dict(FORM_ITEM_DEFINITIONS)

    def extract_items(
        self, content: str, form_type: Union[str, FormType]
    ) -> Dict[str, str]:
        """
        Extract all items from a filing.

        Args:
            content: The filing content (HTML or text)
            form_type: The type of form (e.g., "10-K", "10-Q", "8-K")

        Returns:
            Dictionary mapping item numbers to their content

        Example:
            {
                "1": "Item 1. Business\n...",
                "1A": "Item 1A. Risk Factors\n...",
                "2": "Item 2. Properties\n..."
            }
        """
        # Convert string form type to enum
        if isinstance(form_type, str):
            form_type = self._parse_form_type(form_type)

        if form_type not in self.form_items:
            raise ValueError(f"Unsupported form type: {form_type}")

        # Clean content
        clean_content = self._clean_content(content)

        # Extract table of contents if available
        toc_items = self._extract_table_of_contents(clean_content)

        # Extract items
        items = self._extract_items_from_content(clean_content, form_type, toc_items)

        # Post-process and validate
        return self._post_process_items(items, form_type)

    def extract_specific_items(
        self, content: str, form_type: Union[str, FormType], item_numbers: List[str]
    ) -> Dict[str, str]:
        """
        Extract specific items from a filing.

        Args:
            content: The filing content
            form_type: The type of form
            item_numbers: List of item numbers to extract (e.g., ["1", "1A", "7"])

        Returns:
            Dictionary with only the requested items
        """
        all_items = self.extract_items(content, form_type)
        return {k: v for k, v in all_items.items() if k in item_numbers}

    def _parse_form_type(self, form_type_str: str) -> FormType:
        """Parse string form type to FormType enum."""
        form_type_upper = form_type_str.upper()

        # Handle variations
        if "10-K" in form_type_upper or "10K" in form_type_upper:
            return FormType.FORM_10K
        elif "10-Q" in form_type_upper or "10Q" in form_type_upper:
            return FormType.FORM_10Q
        elif "8-K" in form_type_upper or "8K" in form_type_upper:
            return FormType.FORM_8K
        elif "20-F" in form_type_upper or "20F" in form_type_upper:
            return FormType.FORM_20F
        elif "40-F" in form_type_upper or "40F" in form_type_upper:
            return FormType.FORM_40F
        else:
            raise ValueError(f"Unknown form type: {form_type_str}")

    def _clean_content(self, content: str) -> str:
        """Clean HTML content for better text extraction."""
        # Remove HTML tags but preserve structure
        content = re.sub(r"<[^>]+>", " ", content)

        # Normalize whitespace
        content = re.sub(r"\s+", " ", content)

        # Preserve line breaks for item boundaries
        content = re.sub(
            r"(Item\s+\d+[A-Z]?\.)", r"\n\n\1", content, flags=re.IGNORECASE
        )

        return content.strip()

    def _extract_table_of_contents(self, content: str) -> List[Tuple[str, int]]:
        """
        Extract table of contents to help locate items.

        Returns:
            List of (item_number, position) tuples
        """
        toc_items = []

        # Look for table of contents section
        toc_match = re.search(
            r"TABLE\s+OF\s+CONTENTS(.*?)(?:Item\s+1\.|PART\s+I\s)",
            content,
            re.IGNORECASE | re.DOTALL,
        )

        if toc_match:
            toc_content = toc_match.group(1)

            # Extract item references from TOC
            item_pattern = r"Item\s+(\d+[A-Z]?)\.\s*([^\n\r\.]+)"
            for match in re.finditer(item_pattern, toc_content, re.IGNORECASE):
                item_num = match.group(1).upper()
                toc_items.append((item_num, match.start()))

        return toc_items

    def _extract_items_from_content(
        self, content: str, form_type: FormType, toc_items: List[Tuple[str, int]]
    ) -> Dict[str, ExtractedItem]:
        """Extract items from the main content."""
        items = {}
        item_definitions = self.form_items[form_type]

        # Create patterns for each item
        for item_def in item_definitions:
            # Build pattern with item number and possible variations
            patterns = [
                rf"Item\s+{re.escape(item_def.number)}\.\s*{re.escape(item_def.title)}",
                rf"Item\s+{re.escape(item_def.number)}\.\s*(?=[A-Z])",
                rf"Item\s+{re.escape(item_def.number)}(?:\.|:|\s)",
            ]

            # Add alias patterns
            for alias in item_def.aliases:
                patterns.append(rf"{re.escape(alias)}")

            # Try each pattern
            for pattern in patterns:
                matches = list(
                    re.finditer(pattern, content, re.IGNORECASE | re.MULTILINE)
                )

                if matches:
                    # The same heading occurs in the table of contents, the
                    # cover page, and the body; the body occurrence is the
                    # one followed by real content, so pick the match with
                    # the longest span to the next item heading.
                    best: Tuple[int, int] = (-1, -1)  # (length, start)
                    for m in matches:
                        if toc_items and self._is_in_toc(m.start(), toc_items):
                            continue
                        end = self._find_item_end(content, m.start(), item_definitions)
                        length = end - m.start()
                        if length > best[0]:
                            best = (length, m.start())
                    if best[1] < 0:
                        end = self._find_item_end(
                            content, matches[0].start(), item_definitions
                        )
                        best = (end - matches[0].start(), matches[0].start())

                    start_pos = best[1]
                    end_pos = start_pos + best[0]

                    # Extract content
                    item_content = content[start_pos:end_pos].strip()

                    items[item_def.result_key()] = ExtractedItem(
                        item_number=item_def.result_key(),
                        title=item_def.title,
                        content=item_content,
                        start_position=start_pos,
                        end_position=end_pos,
                    )
                    break

        return items

    def _is_in_toc(self, position: int, toc_items: List[Tuple[str, int]]) -> bool:
        """Check if a position is within the table of contents."""
        if not toc_items:
            return False

        # Rough heuristic: if position is before the last TOC item + buffer
        if toc_items:
            last_toc_pos = max(item[1] for item in toc_items)
            return position < last_toc_pos + 500

        return False

    def _find_item_end(
        self, content: str, start_pos: int, item_definitions: List[ItemDefinition]
    ) -> int:  # noqa: ARG002
        """Find where an item ends (usually the start of the next item)."""
        # Look for the next item
        next_item_pattern = r"Item\s+\d+[A-Z]?[\.:]\s*[A-Z]"

        match = re.search(next_item_pattern, content[start_pos + 10 :], re.IGNORECASE)

        if match:
            return start_pos + 10 + match.start()
        else:
            # No next item found, return end of content
            return len(content)

    def _post_process_items(
        self, items: Dict[str, ExtractedItem], form_type: FormType
    ) -> Dict[str, str]:  # noqa: ARG002
        """Post-process extracted items."""
        processed = {}

        for item_num, extracted_item in items.items():
            # Clean up the content
            content = extracted_item.content

            # Remove excessive whitespace
            content = re.sub(r"\n\s*\n\s*\n", "\n\n", content)

            # Ensure we have some content
            if len(content.strip()) > 50:  # Minimum content threshold
                processed[item_num] = content
            else:
                # Try to handle empty or placeholder items
                if "none" in content.lower() or "not applicable" in content.lower():
                    processed[item_num] = content
                else:
                    processed[item_num] = ""

        return processed

    def get_item_definitions(
        self, form_type: Union[str, FormType]
    ) -> List[ItemDefinition]:
        """Get the item definitions for a specific form type."""
        if isinstance(form_type, str):
            form_type = self._parse_form_type(form_type)

        return self.form_items.get(form_type, [])
